homeassistant/components/fjaraskupan/__init__.py

```diff
@@ -35,21 +35,56 @@
 _LOGGER = logging.getLogger(__name__)
 
 
-@dataclass
-class DeviceState:
-    """Store state of a device."""
+class Coordinator(DataUpdateCoordinator[State]):
+    """Update coordinator for for each device."""
+
+    def __init__(
+        self, hass: HomeAssistant, device: Device, device_info: DeviceInfo
+    ) -> None:
+        """Initialize the coordinator."""
+        self.device = device
+        self.device_info = device_info
+        self._refresh_was_scheduled = False
+
+        super().__init__(
+            hass, _LOGGER, name="Fjäråskupan", update_interval=timedelta(seconds=120)
+        )
+
+    async def _async_refresh(
+        self,
+        log_failures: bool = True,
+        raise_on_auth_failed: bool = False,
+        scheduled: bool = False,
+    ) -> None:
+        self._refresh_was_scheduled = scheduled
+        await super()._async_refresh(
+            log_failures=log_failures,
+            raise_on_auth_failed=raise_on_auth_failed,
+            scheduled=scheduled,
+        )
 
-    device: Device
-    coordinator: DataUpdateCoordinator[State]
-    device_info: DeviceInfo
+    async def _async_update_data(self) -> State:
+        """Handle an explicit update request."""
+        if self._refresh_was_scheduled:
+            raise UpdateFailed("No data received within schedule.")
```

Comment:
I wonder if we should add an `self.async_notify_update_error` to avoid this weird "mark next update as error and then trigger update" approach.

Reply:
A separate timer to keep track of the timeout would be needed then. I don't think this is too weird, apart of the lack of the scheduled parameter to the async_update_data()  function.